core/src/policy.rs

```diff
@@ -0,0 +1,89 @@
+use crate::{
+    changes::ChangeSet,
+    model::{Entity, Model},
+    proto::{CollectionId, NodeId, ID},
+};
+use ankql::ast::Predicate;
+use std::collections::HashSet;
+
+/// The result of a policy check. Currently just Allow/Deny, but will support Trace in the future
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AccessResult {
+    /// The operation is allowed
+    Allow,
+    /// The operation is denied
+    Deny,
+}
+
+impl AccessResult {
+    /// Convenience method to check if access is allowed
+    pub fn is_allowed(&self) -> bool { matches!(self, AccessResult::Allow) }
+}
+
+/// Applications will implement this trait to control access to resources
+/// (Entities and RPC calls) and the Node will be generic over this trait
+pub trait PolicyAgent {
+    /// The context type that will be used for all resource requests.
+    /// This will typically represent a user or service account.
+    type ContextData: ContextData;
+
+    // For checking if a context can access a collection
+    fn can_access_collection(&self, data: &Self::ContextData, collection: &CollectionId) -> AccessResult;
```

Comment:
Thinking through if these need to be a Result.  I suppose it could be up to an implementer to deal with this... but I suspect for some implementations might have errors. I assume an error probably equals deny.  Just curious if we handle that logic, or make the implementer catch all errors themselves and convert to Deny (or whatever).  Just thinking from a security stand point we may want to enforce "If you couldn't answer the question.. then the answer is no").

Just a thought.

Reply:
Yeah, TBH I think the PolicyAgent code is going to change a lot. The main thing this PR achieves is just to break down the type system changes so the Jenga tower doesn't fall over.

I'm going to merge this PR, not because it's ready, but because it's not harmful. This does put the main branch into a non-releasable state, which I don't love, but it's just because its a breaking API change, not because it actually breaks the happy path.